core/arch/aarchxx/mangle.c

```diff
@@ -577,50 +577,212 @@ shrink_reg_for_param(reg_id_t regular, opnd_t arg)
+                PRE(ilist, instr, XINST_CREATE_move(dcontext,
                                                     opnd_create_reg(regparms[i]),
-                                                    arg));
+                                                    args[i]));
+                if (regs[i] != -2)
+                    --usecount[regs[i]];
             }
-        } else {
-            /* FIXME i#1551, i#1569: we only implement naive parameter preparation,
-             * where args are all regs or immeds and do not conflict with param regs.
-             */
-            ASSERT_NOT_IMPLEMENTED(false);
+            regs[i] = i;
+            break;
+        }
+    } while (i < num_regs);
+
+    /* From now on it is safe to use LR as a temporary. */
+
+    /* Set up register arguments that are in cycles. */
+    for (;;) {
+        int first, tmp;
+        for (i = 0; i < num_regs; i++) {
+            if (regs[i] != i)
+                break;
+        }
+        if (i >= num_regs)
+            break;
+        first = i;
+        PRE(ilist, instr, XINST_CREATE_move(dcontext, opnd_create_reg(DR_REG_LR),
+                                            opnd_create_reg(regparms[i])));
+        do {
+            tmp = regs[i];
+            ASSERT(0 <= tmp && tmp < num_regs);
+            PRE(ilist, instr,
+                XINST_CREATE_move(dcontext, opnd_create_reg(regparms[i]),
+                                  tmp == first ? opnd_create_reg(DR_REG_LR) :
+                                  opnd_create_reg(regparms[tmp])));
+            regs[i] = i;
+            i = tmp;
+        } while (tmp != first);
+    }
+
+    /* Set up stack arguments that are (non-zero) constants or SP. */
+    for (i = NUM_REGPARM; i < num_args; i++) {
+        uint off = (i - NUM_REGPARM) * XSP_SZ;
+        opnd_t arg = args[i];
+        if (!opnd_is_reglike(arg)) {
+            if (opnd_is_reg(arg)) {
+                ASSERT(opnd_get_reg(arg) == DR_REG_XSP);
+                insert_get_mcontext_base(dcontext, ilist, instr, DR_REG_LR);
+                PRE(ilist, instr, instr_create_restore_from_dc_via_reg
+                    (dcontext, DR_REG_LR, DR_REG_LR, XSP_OFFSET));
+            } else {
+                ASSERT(opnd_is_immed_int(arg));
+                insert_mov_immed_ptrsz(dcontext, opnd_get_immed_int(arg),
+                                       opnd_create_reg(DR_REG_LR),
+                                       ilist, instr, NULL, NULL);
+            }
+            PRE(ilist, instr, XINST_CREATE_store
+                (dcontext,
+                 opnd_create_base_disp(DR_REG_XSP, DR_REG_NULL, 0, off, OPSZ_PTR),
+                 opnd_create_reg(DR_REG_LR)));
         }
-        DODEBUG({
-            uint j;
-            /* check no reg used by arg conflicts with regparms */
-            for (j = 0; j < i; j++)
-                ASSERT_NOT_IMPLEMENTED(!opnd_uses_reg(args[j], regparms[i]));
-        });
     }
-    return 0;
+
+    return (uint)stack_inc;
```

Comment:
I like it to be a signed type locally because I pass -stack_inc to functions that take an int. Perhaps the return type should be be ptr_int_t or ptr_uint_t: it's an increment to a pointer type and in theory one could have 2^32 arguments on a 64-bit system.